File: telemetry/reservations.py
"""Deterministic reservation model used by tests and sim instrumentation."""

from __future__ import annotations

from dataclasses import dataclass, asdict
from typing import Any


EXCLUSIVE_KINDS = {"chef", "station", "counter", "plate", "stand", "order"}
SOFT_KINDS = {"approach_tile"}


@dataclass
class Reservation:
    id: str
    owner: str
    chef_id: int | None
    kind: str
    resource_id: str
    order_id: int | None
    purpose: str
    created_tick: int
    expires_tick: int
    status: str = "active"
    release_reason: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


class ReservationManager:
    def __init__(self):
        self._reservations: dict[str, Reservation] = {}
        self._next_id = 1

    def _new_id(self) -> str:
        rid = f"r{self._next_id}"
        self._next_id += 1
        return rid

    def active(self) -> list[Reservation]:
        return [r for r in self._reservations.values() if r.status == "active"]

    def conflicts_for(self, kind: str, resource_id: str, ignore_id: str | None = None) -> list[Reservation]:
        conflicts = []
        for reservation in self.active():
            if ignore_id and reservation.id == ignore_id:
                continue
            if reservation.kind == kind and reservation.resource_id == resource_id:
                conflicts.append(reservation)
        return conflicts

    def can_reserve(self, kind: str, resource_id: str) -> bool:
        if kind in SOFT_KINDS:
            return True
        return len(self.conflicts_for(kind, resource_id)) == 0

    def reserve(
        self,
        *,
        owner: str,
        kind: str,
        resource_id: str,
        created_tick: int,
        ttl_ticks: int,
        chef_id: int | None = None,
        order_id: int | None = None,
        purpose: str = "",
    ) -> tuple[Reservation, list[Reservation]]:
        conflicts = self.conflicts_for(kind, resource_id)
        reservation = Reservation(
            id=self._new_id(),
            owner=owner,
            chef_id=chef_id,
            kind=kind,
            resource_id=resource_id,
            order_id=order_id,
            purpose=purpose,
            created_tick=int(created_tick),
            expires_tick=int(created_tick + max(1, ttl_ticks)),
        )
        self._reservations[reservation.id] = reservation
        return reservation, conflicts

    def release(self, reservation_id: str, reason: str = "release") -> Reservation | None:
        reservation = self._reservations.get(reservation_id)
        if reservation and reservation.status == "active":
            reservation.status = "released"
            reservation.release_reason = reason
        return reservation

    def violate(self, reservation_id: str, reason: str = "violation") -> Reservation | None:
        reservation = self._reservations.get(reservation_id)
        if reservation and reservation.status == "active":
            reservation.status = "violated"
            reservation.release_reason = reason
        return reservation

    def expire(self, current_tick: int) -> list[Reservation]:
        expired = []
        for reservation in self.active():
            if reservation.expires_tick <= current_tick:
                reservation.status = "expired"
                reservation.release_reason = "ttl"
                expired.append(reservation)
        return expired

    def snapshot(self) -> dict[str, Any]:
        active = [r.to_dict() for r in self.active()]
        return {
            "active": active,
            "active_count": len(active),
            "all": [r.to_dict() for r in self._reservations.values()],
        }

File: agents/reservations.js
/* Browser-side deterministic reservation manager. */
(function (root) {
  'use strict';

  var HARD = { chef: 1, station: 1, counter: 1, plate: 1, stand: 1, order: 1 };

  function ReservationManager(recorder) {
    this.recorder = recorder || null;
    this.nextId = 1;
    this.items = {};
  }

  ReservationManager.prototype._now = function () {
    var s = root.KitchenAPI && root.KitchenAPI.getState ? root.KitchenAPI.getState() : { time: 0 };
    return { tick: Math.floor(Number(s.time || 0) * 60), ms: performance.now() };
  };

  ReservationManager.prototype.active = function () {
    var out = [];
    Object.keys(this.items).forEach(function (id) {
      if (this.items[id].status === 'active') out.push(this.items[id]);
    }, this);
    return out;
  };

  ReservationManager.prototype.conflicts = function (kind, resourceId) {
    return this.active().filter(function (r) {
      return r.kind === kind && r.resource_id === resourceId;
    });
  };

  ReservationManager.prototype.reserve = function (req) {
    var now = this._now();
    var id = 'r' + this.nextId++;
    var ttl = Math.max(1, req.ttl_ticks || 60 * 30);
    var conflicts = this.conflicts(req.kind, req.resource_id);
    var r = {
      id: id,
      owner: req.owner || 'executor',
      chef_id: req.chef_id == null ? null : req.chef_id,
      kind: req.kind,
      resource_id: String(req.resource_id),
      order_id: req.order_id == null ? null : req.order_id,
      purpose: req.purpose || '',
      created_tick: req.created_tick == null ? now.tick : req.created_tick,
      expires_tick: (req.created_tick == null ? now.tick : req.created_tick) + ttl,
      status: 'active'
    };
    this.items[id] = r;
    if (this.recorder) {
      var data = Object.assign({ conflict: conflicts.length > 0 }, r);
      if (conflicts.length) data.conflicts = conflicts.map(function (c) { return c.id; });
      this.recorder.record('reservation_created', now.tick, now.ms, data);
      if (HARD[req.kind]) {
        conflicts.forEach(function (c) {
          this.recorder.record('reservation_violation', now.tick, now.ms, Object.assign({}, c, {
            reason: 'double_booking',
            conflict_with: id
          }));
        }, this);
      }
    }
    return r;
  };

  ReservationManager.prototype.release = function (id, reason) {
    var r = this.items[id];
    if (!r || r.status !== 'active') return null;
    r.status = 'released';
    r.release_reason = reason || 'release';
    if (this.recorder) {
      var now = this._now();
      this.recorder.record('reservation_released', now.tick, now.ms, r);
    }
    return r;
  };

  ReservationManager.prototype.expire = function (tick) {
    var now = this._now();
    var expired = [];
    tick = tick == null ? now.tick : tick;
    this.active().forEach(function (r) {
      if (r.expires_tick <= tick) {
        r.status = 'expired';
        expired.push(r);
        if (this.recorder) this.recorder.record('reservation_expired', tick, now.ms, r);
      }
    }, this);
    return expired;
  };

  ReservationManager.prototype.snapshot = function () {
    return { active: this.active().map(function (r) { return Object.assign({}, r); }) };
  };

  ReservationManager.prototype.reset = function () {
    this.items = {};
    this.nextId = 1;
  };

  if (typeof module !== 'undefined' && module.exports) module.exports = { ReservationManager: ReservationManager };
  root.ChefReservationManager = ReservationManager;
})(typeof window !== 'undefined' ? window : globalThis);

